### PR TITLE
codegen-llvm: never combine DSOLocal and DllImport

### PR DESCRIPTION
**Pull request by dvdhrm**
Prevent DllImport from being attached to DSOLocal definitions in the LLVM IR. The combination makes no sense, since definitions local to the compilation unit will never be imported from external objects.

Additionally, LLVM will refuse the IR if it encounters the combination (introduced in [1]):

```
  if (GV.hasDLLImportStorageClass())
    Assert(!GV.isDSOLocal(),
           "GlobalValue with DLLImport Storage is dso_local!", &GV);
```

Right now, codegen-llvm will only apply DllImport to constants and rely on call-stubs for functions. Hence, we simply extend the codegen of constants to skip DllImport for any local definitions.

This was discovered when switching the EFI targets to the static relocation model [2]. With this fixed, we can start another attempt at this.

[1] https://smlnj-gitlab.cs.uchicago.edu/manticore/llvm/commit/509132b368efed10bbdad825403f45e9cf1d6e38
[2] https://github.com/rust-lang/rust/issues/101656

### CODE REVIEW

**Comment by petrochenkov** [comment]
Could you add a `src/test/codegen` test case for this situation?
Static relocation mode can be enabled using `// compile-flags` even if it's not the default currently.
@rustbot author

**Comment by petrochenkov** [comment]
>With this fixed, we can start another attempt at this.

Yeah, it would be nice to try reverting https://github.com/rust-lang/rust/pull/101413 when this is merged.

**Comment by dvdhrm** [comment]
@petrochenkov 
> Could you add a `src/test/codegen` test case for this situation?

I now added `static-relocation-mode.rs` to the codegen tests. For now, it *only* tests `dso_local` linkage in the static relocation model.

I would like to also add foreign linkage to this test, but currently *all* declarations are marked `dso_local` when the static relocation model is used. Is that intentional? That is, why can't I write a non-relocatable application which still uses dynamically linked dependencies?

**Comment by petrochenkov** [comment]
@dvdhrm 
>I would like to also add foreign linkage to this test, but currently *all* declarations are marked `dso_local` when the static relocation model is used. Is that intentional? That is, why can't I write a non-relocatable application which still uses dynamically linked dependencies?

Apparently yes, and it's a relatively recent change (https://github.com/rust-lang/rust/pull/83592), but I didn't participate in that work and not sure whether the case you mention was considered.
cc @nagisa

File: src/test/codegen/static-relocation-model.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// CHECK: @extern_static = external dso_local local_unnamed_addr global i8
+// CHECK: define dso_local i8 @access_extern() {{.*}}
+// CHECK: declare dso_local i8 @extern_fn() {{.*}}
+
+#[no_mangle]
+pub fn access_extern() -> u8 {
+    unsafe {
+        extern_decl::extern_fn() + extern_decl::extern_static
+    }
+}
```
Comment (petrochenkov) [inline src/test/codegen/static-relocation-model.rs:25]: I suspect that this test will fail on at least some targets on CI, and you'll have to use some per-target checking like in `src\test\codegen\abi-efiapi.rs`, or limit it to uefi.

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/104791#issuecomment-1325796217

**Comment by dvdhrm** [comment]
CI failed for MacOS, because there the `dso_local`-optimization is actually not as fine-grained, thus `dso_local` is not emitted for the test. I would expect other targets to have similar issues in the future, hence I decided to guard the test on `x86_64-pc-windows-msvc`. Sadly, I couldn't figure out how to run on windows *and* linux targets, but not anything else.

I accidentally also rebased on master, sorry for that! The only change from before is `# only-x86_64-pc-windows-msvc` in the test, and a rename from `static-relocation-model.rs` to `static-relocation-model-msvc.rs`.

@rustbot ready

**Comment by mati865** [comment]
> CI failed for MacOS, because there the `dso_local`-optimization is actually not as fine-grained, thus `dso_local` is not emitted for the test. I would expect other targets to have similar issues in the future, hence I decided to guard the test on `x86_64-pc-windows-msvc`. Sadly, I couldn't figure out how to run on windows _and_ linux targets, but not anything else.
> 
> I accidentally also rebased on master, sorry for that! The only change from before is `# only-x86_64-pc-windows-msvc` in the test, and a rename from `static-relocation-model.rs` to `static-relocation-model-msvc.rs`.
> 
> @rustbot ready

How about `// ignore-macos` maybe combined with `// ignore-ios`?

**Comment by dvdhrm** [comment]
> How about `// ignore-macos` maybe combined with `// ignore-ios`?

Yeah, I thought about this as well, but as I mentioned I expect other targets to possibly have the same shortcuts in the `dso_local`-handling. Those might not be part of CI now, but might get upgraded to tier-2 at any point. I then decided to go with the `only-` marker because the way the `dso_local`-handling in the codegen-module (and llvm, fwiw) is written follows a similar logic. So new targets would get the same behavior as macos, not the same as windows or elf.

Additionally, I would like to test for `dllimport` in there as well, and that will definitely require a msvc-specific test. So maybe the better approach is to copy the test into `static-relocation-model-elf.rs` if we want to test ELF as well?

**Comment by mati865** [comment]
> Additionally, I would like to test for `dllimport` in there as well, and that will definitely require a msvc-specific test.

Seems more like Windows specific (so could be also tested on `windows-gnu`).

> So maybe the better approach is to copy the test into `static-relocation-model-elf.rs` if we want to test ELF as well?

If you think it would be beneficial...

**Pull request merged**